internal/plugin/connectors/tcp/mssql/authentication.go

```diff
@@ -0,0 +1,156 @@
+package mssql
+
+import (
+	"fmt"
+
+	mssql "github.com/denisenkom/go-mssqldb"
+)
+
+func (connector *SingleUseConnector) performPreLoginHandshake() error {
+	// using the default packet size of 4096
+	// (see https://docs.microsoft.com/en-us/sql/database-engine/configure-windows/configure-the-network-packet-size-server-configuration-option)
+	clientBuffer := mssql.NewTdsBuffer(4096, connector.clientConn)
+	preloginRequest, err := mssql.ReadPreloginWithPacketType(clientBuffer, mssql.PackPrelogin)
+	if err != nil {
+		return fmt.Errorf("failed to read prelogin request: %s", err)
+	}
+
+	preloginResponse := preparePreloginResponse(preloginRequest)
+
+	err = mssql.WritePreloginWithPacketType(clientBuffer, preloginResponse, mssql.PackReply)
+	if err != nil {
+		return fmt.Errorf("failed to write prelogin response: %s", err)
+	}
+
+	// we actually don't need the client's handshake response.
+	// we just need for them to not be blocked
+	err = clientBuffer.ReadNextPacket()
+	if err != nil {
+		return fmt.Errorf("failed to read client login message: %s", err)
+	}
+
+	return nil
+}
+
+// The prelogin response message is identical to the request with minor changes
+func preparePreloginResponse(preLoginRequest map[uint8][]byte) map[uint8][]byte {
+	preloginResponse := preLoginRequest
+
+	// According to the mssql docs, The client can use the VERSION returned from
+	// the server to determine which features SHOULD be enabled or disabled. The
+	// client SHOULD do this only if it can determine whether a feature is
+	// supported by the current version of the database.
+	//
+	// In our case, the client will not use the VERSION so we don't need the actual
+	// version of the server
+	preloginResponse[mssql.PreloginVERSION] = []byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
```

Comment:
As you mentioned this version is necessary fr the client to determine which features are available on the server. We need to dutifully transmit whatever comes from the server.

When I run `./start` in the test folder and wait for mssql to come up, then run:
```
sqlcmd -S "127.0.0.1,2223" -U x -P "x"
```

It results in:
```
Sqlcmd: Error: Microsoft ODBC Driver 17 for SQL Server : ODBC Driver 17 for SQL Server does not support connections to SQL Server 2000 or earlier versions..
Sqlcmd: Error: Microsoft ODBC Driver 17 for SQL Server : Client unable to establish connection.
```

This is 100% a result of PreloginVERSION. Connecting directly works fine.

Reply:
very nice catch